internal/reminder/reminder.go

```diff
@@ -0,0 +1,197 @@
+	level := util.ViperLogLevelToZerologLevel(config.LoggingLevel)
+	r := &reminder{
+		store:          store,
+		cfg:            config,
+		stop:           make(chan struct{}),
+		repoListCursor: make(map[projectProviderPair]string),
+		logger:         zerolog.New(os.Stdout).Level(level).With().Timestamp().Logger(),
+	}
+	err := r.restoreCursorState()
+	if err != nil {
+		// Non-fatal error, if we can't restore the cursor state, we'll start from scratch.
+		r.logger.Error().Err(err).Msg("error restoring cursor state")
+	}
+
+	pub, cl, err := r.setupSQLPublisher(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	r.eventPublisher = pub
+	r.eventDBCloser = cl
+	return r, nil
+}
+
+// Start starts the reminder by sending reminders at regular intervals
+func (r *reminder) Start(ctx context.Context) error {
+	select {
+	case <-r.stop:
+		return errors.New("reminder stopped, cannot start again")
+	default:
+	}
+
+	interval, err := time.ParseDuration(r.cfg.RecurrenceConfig.Interval)
+	if err != nil {
+		return err
+	}
+
+	if interval <= 0 {
+		return fmt.Errorf("invalid interval: %s", r.cfg.RecurrenceConfig.Interval)
+	}
+
+	r.ticker = time.NewTicker(interval)
+	defer r.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.stop:
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.ticker.C:
+			// In-case sending reminders i.e. iterating over entities consumes more time than the
+			// interval, the ticker will adjust the time interval or drop ticks to make up for
+			// slow receivers.
+			if errs := r.sendReminders(ctx); errs != nil {
+				for _, err := range errs {
+					r.logger.Error().Err(err).Msg("reconciliation request unsuccessful")
+				}
+			}
+		}
+	}
+}
+
+// Stop stops the reminder service
+func (r *reminder) Stop() {
+	if r.ticker != nil {
+		defer r.ticker.Stop()
+	}
+	r.stopOnce.Do(func() {
+		close(r.stop)
+		r.eventDBCloser()
+	})
+}
+
+func (r *reminder) storeCursorState() error {
+	r.logger.Debug().Msg("storing cursor state")
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+
+	data := map[string]interface{}{
+		"projectListCursor": r.projectListCursor,
+		"repoListCursor":    r.repoListCursor,
+	}
+	if err := enc.Encode(data); err != nil {
+		return err
+	}
+
+	return os.WriteFile(r.cfg.CursorFile, buf.Bytes(), 0600)
+}
+
+func (r *reminder) restoreCursorState() error {
+	if _, err := os.Stat(r.cfg.CursorFile); os.IsNotExist(err) {
+		return nil
+	}
+
+	data, err := os.ReadFile(r.cfg.CursorFile)
+	if err != nil {
+		return err
+	}
+
+	buf := bytes.NewBuffer(data)
+	dec := gob.NewDecoder(buf)
+
+	cursorData := make(map[string]interface{})
+
+	if err := dec.Decode(&cursorData); err != nil {
+		return err
+	}
+
+	r.projectListCursor = cursorData["projectListCursor"].(string)
+	r.repoListCursor = cursorData["repoListCursor"].(map[projectProviderPair]string)
```

Comment:
Do you want to use the 2-value return form to avoid a panic if the key is missing?

Reply:
Done.